### PR TITLE
refactor: flatten src directory

### PR DESCRIPTION
**Pull request by jacobdadams**
I propose we flatten the `src/project-name` subdirectory by removing the `project-name` and putting everything in `src` by default. This shortens submodule import statements, keeps smaller programs simpler, and simplifies putting a `__main__.py` in the root for direct CLI invocation.

If larger, well-engineered programs need to organize submodules into directories, they can still do that.

For example, let's start with the following structure:
```
foo-project
├─ __main__.py
├─ __init__.py
├─ prime_class.py
├─ subprime1.py
└─ subprime2.py
```

`__main__.py` contains the CLI code and instantiates an instance of `prime_class`. This allows us to call the program straight from the CLI w/o .py: `python c:\git\foo-project`. As part of it's work, `prime_class` instantiates multiple instances of both `subprime1` and `subprime2`.

If we try to modify it thus:
```
foo-project
├─ src
|    └─ foo-project
|       ├─ __init__.py
|       ├─ prime_class.py
|       ├─ subprime1.py
|       └─ subprime2.py
└─ __main__.py
```
 To import anything into `__main__.py`, we then have to add a `__init__.py` to the root of `src` and import as thus: `from  src.foo-project.prime_class import prime` and `from src.foo-project.subprime1 import sub1`

The double gotcha is that python modules can't have a `-` in their name, contrary to our project/repo naming convention. `src/foo-project` would need to be renamed `src/foo_project`, creating an inconsistency with the repo name. (Perhaps it's better if we change our naming convention too, at least for python projects?) 

Flattening `src` gives us the following structure:

```
foo-project
├─ src
|   ├─ __init__.py
|   ├─ prime_class.py
|   ├─ subprime1.py
|   └─ subprime2.py
└─ __main__.py
```

### CODE REVIEW

**Comment by stdavis** [comment]
I'm all for simplification but I think that this might have implications when you pip install a project and possibly for running tests (see the travis build output). I think that the idea is that you shouldn't be importing stuff like this: 
```py
from projectfolder.src.packagename import somemodule
```

Rather you should be installing it and then importing like this:
```py
from packagename import somemodule`
```

To run the command line during development you would do something like this from `src`:

```
python -m packagename
```

Your `__main__.py` file should go in `src/packagename`.

I feel like @steveoh had some other reasons for the `src` structure by not sure.

With all of this being said, maybe this project structure is a bit much for a simple script? Maybe we should have something like `agrc/python-package` and `agrc/python-script` as templates?

**Comment by jacobdadams** [comment]
Ok, I see why I couldn't get this to work as outlined in the template- I'm not installing it with setup.py, and I'm trying to avoid having to cd into src to run the program. Part of me feels like those are too complicated for a simple script, but that could also be me being lazy :) (plus I've not seen a popular python package that requires you to cd into the src directory to run it from the CLI)

As I type that, maybe what I'm running up against is the possible dichotomy of a python project that is both a module and a CLI program. Is it better engineering to separate that into two projects- a module/library project and an interface project? Something like `agol-validator-pallet` and `agol-validator`, where `a-v-p` is run through the forklift framework, and `a-v` is a simple doc-opt CLI that calls `a-v-p`?

**Comment by jacobdadams** [comment]
Also, I am in favor of creating a more simple template for scripts. I think the beauty and challenge of python is that it allows you to both create quick, simple scripts and engineer complex programs. 

Because of this, I think many people, including myself until recently, only do simple scripts and make the judgement call that a quick and dirty script that works for their one task is a better use of their time than learning the ins-and-outs of software engineering. Trying to constrain this type of coder to the full framework of setup, test coverage, and complex object-oriented design can create the frustration that leads to them just completely ignoring the template. 

To stretch an analogy, Netwon's laws and the derived kinematics equations get you close enough to do lots of things in the physical world with a basic understanding of calculus, or even just algebra. However, to be really precise, you need lots of calculus and general relativity or quantum mechanics. We don't need calculus to build a footbridge over a stream, and it also doesn't hurt to make it look good and be more stable than a couple of 2x4s lashed together with duct tape (apologies to any Red Green Show fans out there).

**Comment by steveoh** [comment]
> Ok, I see why I couldn't get this to work as outlined in the template- I'm not installing it with setup.py, and I'm trying to avoid having to cd into src to run the program. Part of me feels like those are too complicated for a simple script, but that could also be me being lazy :) (plus I've not seen a popular python package that requires you to cd into the src directory to run it from the CLI)

If you pip install the project you can reference it from anywhere with the module name. The `python -m projectname` is another way to do this and is typically used during development but I do not think is necessary with a setup.py.

> As I type that, maybe what I'm running up against is the possible dichotomy of a python project that is both a module and a CLI program. Is it better engineering to separate that into two projects- a module/library project and an interface project? Something like `agol-validator-pallet` and `agol-validator`, where `a-v-p` is run through the forklift framework, and `a-v` is a simple doc-opt CLI that calls `a-v-p`?

I do not think it is required to separate those into different projects but most likely as separate files. As long as the module is able to be imported everything should be fine.

We could all pair together if you wanted to come up with a working solution.

> Trying to constrain this type of coder to the full framework of setup, test coverage, and complex object-oriented design can create the frustration that leads to them just completely ignoring the template.

I don't think that this project structure forces you to understand software engineering any more than writing a quick and dirty script. There is no complex object oriented design in the template; There is not even a class. Please understand there is no rule that every single python thing AGRC writes needs to use this template.

I believe the template is a tool that can enable us to improve as code authors. It is easy to write a test if I want to. It is easy to distribute the module with others if I want to. I'd rather we make these scenarios simple to take advantage of or ignore rather than the other way around.

> I have been frustrated in the past that I have to jump through all these hoops I don't understand and didn't see as necessary just to get something that you guys won't pick to pieces in a code review. To me, something that worked for my problem at that moment was more important than a beautiful, well-integrated and test script.

Scott and I are always happy and willing to walk through things to help people understand and improve. We aren't being malicious in our code reviews and we definitely do not have intentions to demean people. 

I know there will be frustrations with technology especially when we know there is a simpler or more familiar route. We are all trying to do our best while getting things done. Quick and dirty is necessary sometimes. Sometimes it is only quick and dirty in the beginning - deferring the extra time to the future.

**Comment by stdavis** [comment]
Coming from the perspective of someone who is not a cs-degree, certified code monkey, I can understand the pain of trying to "train up" to @steveoh's level. I can totally relate to feeling like "my solution was working fine and more simple, why are you asking me to make it more complicated"! However, I've learned through sad experience that it's usually better to go through the pain of conforming to industry standards and best practices. Or at least making that your goal. I've got projects that are definitely not up to snuff. But I try to make them better each time that I touch them.

In short, I feel your pain and just wanted to say that it's worth it.
